src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -36,19 +38,107 @@ internal static bool IsEnabled()
             //from DiagnosticListener right after the check. So some requests happening right after subscription starts
             //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
 
-            Guid loggingRequestId = Guid.NewGuid();
-            LogHttpRequest(request,  loggingRequestId);
+            Task<HttpResponseMessage> responseTask;
 
-            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
-
-            LogHttpResponse(responseTask, loggingRequestId);
+            //We need to know if activity events are enabled and send Request/Response events otherwise
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName))
+            {
+                if (Activity.Current != null)
+                {
+                    responseTask = SendInstrumentedAsync(request, cancellationToken);
+                }
+                else
+                {
+                    //null Activity.Current means that incoming request was not instrumented
+                    //and there is nothing we can do for outgoing request. 
+                    //Activity events are enabled so we don't send Request/Response events
+                    responseTask = base.SendAsync(request, cancellationToken);
+                }
+            }
+            else 
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                LogHttpRequest(request, loggingRequestId);
+                responseTask = base.SendAsync(request, cancellationToken);
+                LogHttpResponse(responseTask, loggingRequestId);
+            }
             return responseTask;
         }
 
         #region private
 
-        private static readonly DiagnosticListener s_diagnosticListener =
-            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+        private async Task<HttpResponseMessage> SendInstrumentedAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Activity activity = null;
+            //check if user wants THIS request to be instrumented
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName, request))
+            {
+                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
+
+                s_diagnosticListener.StartActivity(activity, new { Request = request });
+
+                //inject headers unless user does not want it to happen
+                //we may pass request to IsEnabled, but not clear how useful it is
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityPropagateName))
+                {
+                    request.Headers.Add(HttpKnownHeaderNames.RequestId, activity.Id);
+                    
```

Comment:
Instrumentation is opt-in. To enable it, you need to listen to HttpHanlerDiagnosticListener and enable System.Net.Http.Activity event.

There is already an API for custom implementation - DelegatingHandler. Users can track outgoing HTTP requests there, if default Activity instrumentation does not fit. Custom DelegatingHandler does not need to repeat all this logic, since it can directly write to the logging system and do filtering in place. 

Please also check @vancem reply:

>@cwe1ss -on the issue on wiring in the names RequestId and CorrelationContext into the code.
>
>We do care about interoperability and are not discounting your concern. Our basic plan is
>
> * There should be a worldwide standard for this (frankly OpenTracing SHOULD have picked one). If there HAD been an RFC or even if OpenTracing had suggested something we would have chosen it and we would not be having this conversation.
> * Because of the lack of standards we FULLY EXPECT to have to add logic that allows flexibility at the HTTP boundary. But this should be thought of as a 'compatibility hack' until everyone gets on the standard (you could imagine removing this code eventually because it is not useful in the presence of a strong standard).
> * There seems to be little value in making an extensibility mechanism so that we can 'pass the problem off' to something outside of HTTP code. After all arguably information about what goes in HTTP headers and their format something that belongs in HTTP code.
> * Ideally we would only have to add this compatibility flexibility ON INCOMMPING Http. The reason is that it not inefficient to check for N different synonyms for a HTTP header, but it IS inefficient to have to write the fields N different ways because you are using N different logging systems that have different conventions. Our plan is to have an environment variable that allows you to map 'foreign' names to their 'standard' one. If all logging systems do this, then no one need sot fix their OUTGOING http. We will see...
>
> However our first order of business is to get SOMETHING up and running. Certainly by getting this minimum we enable the scenario where you only use .NET code (a nice start), and even cross-logging cases if their loggers are sufficiently flexible (they have the incoming mapping as we suggest in #4). When we get this working, we will add #4 and things 'should' work. We will see.
>
> So we ARE interested in getting cross-logging systems working. You could be very helpful in making that actually work (and ideally we get other logging systems to be adaptable (or even come with us as we propose and adopt a standard)). We really would have RATHER simply followed someone else's lead, but given that that was not there, we want to 'fix it right' (which is with a standard). We don't have to control the standard, it just needs to exist.